draft/python/_py-test.yaml

```diff
@@ -0,0 +1,292 @@
+# SPDX-FileCopyrightText: Copyright 2024, the RCSB PDB, University of California, Rutgers University, and contributors
+# SPDX-FileCopyrightText: Copyright 2020-2024, Contributors to CICD
+# SPDX-PackageHomePage: https://github.com/dmyersturnbull/cicd
+# SPDX-License-Identifier: Apache-2.0
+# SPDX-FileCopyrightText: Copyright 2020-2023, Contributors to Tyrannosaurus
+# SPDX-PackageHomePage: https://github.com/dmyersturnbull/tyrannosaurus
+# SPDX-License-Identifier: Apache-2.0
+
+# This workflow is designed to run various types of tests and analyses on the codebase.
+# It is triggered from other workflows using `workflow_call`, with specific ref inputs.
+# Different jobs cover tests across multiple operating systems (Ubuntu, Windows, macOS) and Python versions.
+# Some jobs are conditional, triggered only by specific comments on pull requests or specific events,
+# such as '/please-review', '/please-run-tests', '/please-check-docker-build', etc.
+# TODO: Allow tests after failure to run as long as not cancelled
+
+name: Test
+run-name: >
+  ${{ github.workflow }} \
+  ${{ github.ref_name }} \
+  (triggered by ${{ github.ref_type }})
+
+on:
+  workflow_call:
+    inputs:
+      ref:
+        type: string
+        description: Ref to check out
+        default: main
+
+permissions:
+  contents: read
+  statuses: write
+  actions: read
+  security-events: write
+
+jobs:
+
+  preliminary:
+    name: Start Markdown table
+    runs-on: ubuntu-latest
+    steps:
+      - run:
+          echo "\n| Test | Outcome |\n| --- | --- |" >> ${GITHUB_STEP_SUMMARY}
+
+  pytest:
+    # This job runs unit tests on multiple operating systems (Ubuntu, Windows, macOS) with Python 3.12.
+    # It is triggered by pull requests, specific comments ('/please-review', '/please-run-tests'),
+    # or when called directly from another workflow.
+    # The job uploads coverage reports as artifacts for each combination of OS and Python version.
+    if: >
+      github.event_name == 'pull_request' ||
+      github.event.comment.body == '/please-review' ||
+      github.event.comment.body == '/please-run-tests' ||
+      github.event_name == 'workflow_call'
+    strategy:
+      matrix:
+        os:
+          - ubuntu-latest
+          - windows-latest
+          - macos-latest
+        python:
+          - '3.12'
+    name: Run tests for ${{ matrix.os }}, Python ${{ matrix.python }}
+    runs-on: ${{ matrix.os }}
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.inputs.ref || github.ref }}
+      - uses: actions/setup-python@v5
+        with:
+          python-version: ${{ matrix.python }}
+      - name: Install dependencies and run tests
+        id: test
+        run: |
+          pip install hatch~=1.12
```

Comment:
Checking I understand. Currently, multiple `_test.yaml` jobs install Python, build the main package (let's say `pkg`) and download and install `pkg`'s dependencies along with some extra packages:

- `pytest` (builds `pkg`; installs dependencies, Hatch, and test frameworks)
- `docker` (builds `pkg`; installs dependencies and Hatch)
- `docs` (installs Hatch and doc-building tools)
- `security` and `codeql` also set up Python

The `pytest` and `docker` redundancy isn't great. We could run tests inside Docker to avoid it. The downside is with testing non-containerized RCSB Python packages (i.e. libraries, clients, etc.): Currently, excluding a Dockerfile disables that step, while tests are still run. If we move tests to Docker, things are less straightforward.

As for having to build separately for all 4 jobs -- if I understand correctly, you're proposing a solution for this?

Reply:
`hatch`, test frameworks, and doc-building tools would be installed as part of the Python runner image. When the pipeline runs for `pkg`, it'll pull and use the Python runner image and avoid having to download and install those common build tools. It'll just have to install the `pkg` specific dependencies and build `pkg`.